### PR TITLE
96 migrate to typescript

### PR DESCRIPTION
**Pull request by chinchaun**
first steps

### CODE REVIEW

**Comment by pranaygp** [comment]
Thanks so much @chinchaun. I'll review this tomorrow :)

**Comment by chinchaun** [comment]
I put 2 TODO with comments that you may wanna look at it

File: packages/core/lib/scope.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  return resolveModule(name, scope.parent);
+};
+
+/**
+ * Recursively find a rule matching the name by walking up the scope/environment inheritance chain
+ * @param {string | Word} name The rule name or Word AST node
+ * @param {Scope} scope The scope to use
+ */
+const resolveRule = (name: any , scope: IScope) : any => {
+  // TODO:  name.name.name? dafuq, i need some explanation for this
```
Comment (pranaygp) [inline packages/core/lib/scope.ts:32]: lol my bad

the AST is gross and we can improve the naming, but it has to do with:
https://github.com/windsorio/browse/blob/master/packages/parser/lib/index.js#L325-L353

the structure is InitRule -> identifier -> word -> (a string), and each AST node in that chain has a property called `name` which points to the child

**Review by pranaygp — COMMENTED** [review]
Lot of `any`. I think this is good to ship if you wanna create an issue and do that in a new PR, or we can just do it in this PR

File: packages/parser/lib/interfaces/ILiteralCreator.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import QuoteTypes from "../types/QuoteTypes";
+
+export default interface ILiteralCreator {
```
Comment (pranaygp) [inline packages/parser/lib/interfaces/ILiteralCreator.ts:3]: what's the difference between `interface` and `type` in ts. i know i tend to use `type` and the differences are subtle, but just curious because i don't know when to use one of the other.

also, would it  be possible to have the creator interfaces and the actual type (here: `ILiteralCreator` and `Literal`) in the same ts file since the types are so similar, it'd be nice to only have to update one file instead of 2

File: packages/parser/lib/interfaces/IASTNode.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import EvalExprTypeEnum from "../../../core/lib/enums/EvalExprTypeEnum";
+
+export default interface IASTNode {
+  type: EvalExprTypeEnum,
```
Comment (pranaygp) [inline packages/parser/lib/interfaces/IASTNode.ts:4]: let's rename`EvalExprtypeEnum` to `ASTTypeEnum`

File: packages/parser/lib/interfaces/IASTNode.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import EvalExprTypeEnum from "../../../core/lib/enums/EvalExprTypeEnum";
```
Comment (pranaygp) [inline packages/parser/lib/interfaces/IASTNode.ts:1]: we can't import cross package like this

File: packages/parser/lib/ast.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Utility functions to create AST nodes
+ */
+import Literal from "./types/Literal";
+import EvalExprTypeEnum from "../../core/lib/enums/EvalExprTypeEnum";
```
Comment (pranaygp) [inline packages/parser/lib/ast.ts:5]: we can't import cross package like this

File: packages/parser/lib/interfaces/IASTNode.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+
+export default interface IASTNode {
+  type: EvalExprTypeEnum,
+  source: any
```
Comment (pranaygp) [inline packages/parser/lib/interfaces/IASTNode.ts:5]: `any` is fine, but it'd be really nice to have ohm's `source`

see
1. https://github.com/harc/ohm/blob/099bd7861d7493d5660dd2b034020643077399cb/ohm-js/index.d.ts#L262-L265
2. https://github.com/harc/ohm/blob/099bd7861d7493d5660dd2b034020643077399cb/ohm-js/index.d.ts#L304-L351

File: packages/parser/lib/ast.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+import ILiteralCreator from "./interfaces/ILiteralCreator";
+
+export function literal(params: ILiteralCreator): Literal {
```
Comment (pranaygp) [inline packages/parser/lib/ast.ts:8]: I like this `ILiteralCreator` idea 😃

File: packages/parser/lib/types/Literal.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import QuoteTypes from "./QuoteTypes";
+import EvalExprTypeEnum from "../../../core/lib/enums/EvalExprTypeEnum";
```
Comment (pranaygp) [inline packages/parser/lib/types/Literal.ts:2]: ❌ cross package import

File: packages/parser/lib/types/QuoteTypes.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+type QuoteTypes = '"' | "'" | '`' | '|';
```
Comment (pranaygp) [inline packages/parser/lib/types/QuoteTypes.ts:1]: oh, `""` (empty string) is also a valid `quoteType` (for implicit strings)

File: package.json [hunk 10]
```diff
@@ -20,6 +20,7 @@
     }
   },
   "dependencies": {
+    "@types/node": "^14.11.2",
```
[the remark this answers is not in the document]
Comment (chinchaun) [reply package.json:23]: my bad

File: packages/parser/lib/interfaces/ILiteralCreator.ts (hunk 2 above)
Comment (chinchaun) [reply packages/parser/lib/interfaces/ILiteralCreator.ts:3]: For example, here the `interface` is been used as a contract in the case, send me anything that you want I don't care as long as has these things and the `type` is using for shaping things and giving them names, in the form of this object with these properties it's called `Literal`, let me know if that makes sense.

I think the ILiteralCreator is going to be a more generic thing in the future that why is in another file, if that doesn't happen it can be moved for sure.

File: packages/core/lib/index.ts [hunk 11]
```diff
@@ -0,0 +1,349 @@
...
+// TODO: having a global moduleCache doesn't feel good
+const moduleCache = new Map();
+
+
+const getNewScope = (parent?: any) : IScope => {
+  if (!parent) {
+    parent = getSTD({ evalRule, evalRuleSet, getNewScope });
+  }
+
+  // TODO: scope shouldn't be a class ?
```
[the remark this answers is not in the document]
Comment (chinchaun) [reply packages/core/lib/index.ts:26]: Ok I understand the pro/cons that you mention, so I think the best approach it's to give a type to the scope but I need to dig a little bit more to see if there is a case for the interface IScope and also how are differences between the ohm scope and browse Scope

File: packages/core/lib/index.ts [hunk 12]
```diff
@@ -0,0 +1,349 @@
+"use strict";
+
+import fs from "fs";
+import path from "path";
+
+const parser = require("@browselang/parser");
```
[the remark this answers is not in the document]
Comment (chinchaun) [reply packages/core/lib/index.ts:6]: util parser it's fully ts, this is the way 😝

File: packages/parser/lib/interfaces/IASTNode.ts (hunk 4 above)
Comment (chinchaun) [reply packages/parser/lib/interfaces/IASTNode.ts]: yup, I'm going to change this when is fully ts

File: packages/core/lib/enums/ASTTypeEnum.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+enum ASTTypeEnum {
```
Comment (pranaygp) [inline packages/core/lib/enums/ASTTypeEnum.ts:1]: we need to share this between both packages, so it can't be in core

File: packages/core/lib/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (pranaygp) [reply packages/core/lib/index.ts:26]: Sure! Feel free to ask me questions or if you want a deeper explanation over slack, but in short - the scope object represents the concept of scoping in the browse language.

read this: https://craftinginterpreters.com/statements-and-state.html#scope

tl;dr
```js
// In javascript
let x = 12;
if (true) {
    // the '{' and  '}' creates a new scope
    let x = 13; // this `x` shadows the `x` in the higher scope
    console.log(x) // => 13
}
console.log(x) // => 12
```

File: packages/parser/lib/ast.ts [hunk 14]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Utility functions to create AST nodes
+ */
+import Literal from "./types/Literal";
+import ASTTypeEnum from "../../core/lib/enums/ASTTypeEnum";
```
Comment (pranaygp) [inline packages/parser/lib/ast.ts:5]: can't have a cross package import

File: packages/parser/lib/interfaces/IASTNode.ts [hunk 15]
```diff
@@ -0,0 +1,8 @@
+import ASTTypeEnum from "../../../core/lib/enums/ASTTypeEnum";
```
Comment (pranaygp) [inline packages/parser/lib/interfaces/IASTNode.ts:1]: can't have a cross package import

File: packages/parser/lib/interfaces/IASTNode.ts [hunk 16]
```diff
@@ -0,0 +1,8 @@
...
+
+import { Interval } from "ohm-js";
+
+export default interface IASTNode {
+  type: ASTTypeEnum,
+  source: Interval
```
Comment (pranaygp) [inline packages/parser/lib/interfaces/IASTNode.ts:7]: 😍

File: packages/parser/lib/types/Literal.ts [hunk 17]
```diff
@@ -0,0 +1,13 @@
+import QuoteTypes from "./QuoteTypes";
+import ASTTypeEnum from "../../../core/lib/enums/ASTTypeEnum";
```
Comment (pranaygp) [inline packages/parser/lib/types/Literal.ts:2]: can't have a cross package import

File: packages/core/lib/enums/ASTTypeEnum.ts (hunk 13 above)
Comment (chinchaun) [reply packages/core/lib/enums/ASTTypeEnum.ts]: moved the type to the new shared package

File: packages/parser/lib/ast.ts (hunk 14 above)
Comment (chinchaun) [reply packages/parser/lib/ast.ts]: fixed

File: packages/parser/lib/interfaces/IASTNode.ts (hunk 15 above)
Comment (chinchaun) [reply packages/parser/lib/interfaces/IASTNode.ts]: fixed

File: packages/core/stdlib/datetime/native.ts [hunk 18]
```diff
@@ -1,14 +1,11 @@
+import IScope from "../../lib/interfaces/IScope";
+
 const isNullish = (v) => v === undefined || v === null;
+// TODO: isNullish is also declared on the utils file, what are the difference between them and why the are named the same
```
Comment (pranaygp) [inline packages/core/stdlib/datetime/native.ts:4]: oh, both of them are the same. I just forgot to use it from the utils file. We can actually just import that instead of rewriting it here

File: packages/core/lib/utils.ts [hunk 19]
```diff
@@ -0,0 +1,192 @@
...
+    return "nil";
+  }
+  if (typeof jsValue === "function") {
+    return "rule ...";
+  }
+
+  const indent = new Array(depth + 1).join("  "); // indent
+  // Arrays
+  if (Array.isArray(jsValue)) {
+    // TODO: if it's and array this should be lenght instead of size?
```
Comment (pranaygp) [inline packages/core/lib/utils.ts:33]: yup! good catch. that's a bug 🐛
